rotkehlchen/balances/historical.py

```diff
@@ -120,6 +121,67 @@ def get_assets_amounts(
 
         return amounts, last_event_id
 
+    def get_historical_netvalue(
+            self,
+            from_ts: Timestamp,
+            to_ts: Timestamp,
+    ) -> tuple[dict[Timestamp, FVal], list[tuple[str, Timestamp]]]:
+        """Calculates historical net worth per day within the given time range.
+
+        Uses asset balances per day combined with historical prices to calculate
+        the total net worth in the user's main currency for each day.
+
+        May raise:
+            - NotFoundError if no events exist in the specified time period.
+
+        Returns:
+            - A mapping of timestamps to net worth in main currency for each day
+            - A list of (asset id, timestamp) tuples where price data was missing
+        """
+        events, main_currency = self._get_events_and_currency(from_ts=from_ts, to_ts=to_ts)
+        if len(events) == 0:
+            raise NotFoundError(f'No historical data found within {from_ts=} and {to_ts=}.')
+
+        current_balances: dict[Asset, FVal] = defaultdict(FVal)
+        daily_balances: dict[Timestamp, dict[Asset, FVal]] = defaultdict(lambda: defaultdict(FVal))
+        for event in events:
+            if (
+                (direction := event.maybe_get_direction()) is None or
+                direction == EventDirection.NEUTRAL
+            ):
+                continue
+
+            if direction == EventDirection.IN:
+                current_balances[event.asset] += event.balance.amount
+            elif direction == EventDirection.OUT:
+                current_balances[event.asset] -= event.balance.amount
+
+            daily_balances[timestamp_to_daystart_timestamp(ts_ms_to_sec(event.timestamp))] = current_balances.copy()  # noqa: E501
```

Comment:
you are copying a dict that can be big several times. Can't you keep track of the last day, current day and 

```
if current_day < this_event_day:
     daily_balances[timestamp_to_daystart_timestamp(ts_ms_to_sec(event.timestamp))] = current_balances.copy()
```

you don't need to copy always, only when you change from one day to the next

Reply:
i didn't find a clean way to do this and trying different variants resulted in wrong calculations.
```py
 timestamp = ts_ms_to_sec(event.timestamp) if isinstance(event, HistoryBaseEntry) else event.timestamp  # noqa: E501
  current_day = timestamp_to_daystart_timestamp(timestamp)
  if last_day is not None and current_day > last_day:
      daily_balances[last_day] = current_balances.copy()

  last_day = current_day

if last_day is not None:
  daily_balances[last_day] = current_balances.copy()
```